nflows/transforms/standard.py

```diff
@@ -1,82 +1,92 @@
 """Implementations of some standard transforms."""
 
+from typing import Optional
+import warnings
+import numpy as np
 import torch
-
+from nflows.utils.torchutils import ensure_tensor
 from nflows.transforms.base import Transform
 
 
 class IdentityTransform(Transform):
     """Transform that leaves input unchanged."""
 
-    def forward(self, inputs, context=None):
-        batch_size = inputs.shape[0]
-        logabsdet = torch.zeros(batch_size)
+    def forward(self, inputs: torch.Tensor, context=Optional[torch.Tensor]):
+        num_batches = inputs.shape[0]
+        logabsdet = torch.zeros(num_batches)
         return inputs, logabsdet
 
-    def inverse(self, inputs, context=None):
+    def inverse(self, inputs: torch.Tensor, context=Optional[torch.Tensor]):
         return self(inputs, context)
 
 
-class AffineScalarTransform(Transform):
-    """Computes X = X * scale + shift, where scale and shift are scalars, and scale is non-zero."""
-
-    def __init__(self, shift=None, scale=None):
+class PointwiseAffineTransform(Transform):
+    def __init__(
+        self,
+        shift: torch.Tensor = torch.tensor(0.0),
+        scale: torch.Tensor = torch.tensor(1.0),
+    ):
         super().__init__()
+        shift, scale = map(ensure_tensor, (shift, scale))
 
-        if shift is None and scale is None:
-            raise ValueError("At least one of scale and shift must be provided.")
-        if scale == 0.0:
-            raise ValueError("Scale cannot be zero.")
+        # reject scales < ~0 (upto dtype precision)
+        is_scale_positive = torch.isfinite(torch.log(scale)).any()
+        if not is_scale_positive:
+            raise ValueError("Scale must be strictly positive.")
 
-        self.register_buffer(
-            "_shift", torch.tensor(shift if (shift is not None) else 0.0)
-        )
-        self.register_buffer(
-            "_scale", torch.tensor(scale if (scale is not None) else 1.0)
-        )
+        self.register_buffer("_shift", shift)
+        self.register_buffer("_scale", scale)
 
     @property
     def _log_scale(self):
-        return torch.log(torch.abs(self._scale))
-
-    def forward(self, inputs, context=None):
-        batch_size = inputs.shape[0]
-        num_dims = torch.prod(torch.tensor(inputs.shape[1:]), dtype=torch.float)
+        return torch.log(self._scale)
+
+    # XXX memoize?
+    def _batch_logabsdet(self, batch_shape: torch.Size):
+        """Return log abs det with input batch shape."""
+        scale_numel = int(np.prod(self._scale.shape))
+        if scale_numel > 1:
+            return self._log_scale.expand(batch_shape).sum()
+        else:
+            # optimise for scalar scale: expand = numel equal scalars
+            # also needed to match test at eps < 1E-5
```

Comment:
I think the comment could be more informative than "This is to make the test pass": computing `x*y` is more numerically accurate than computing the same thing by adding `x` `y` times: we lose precision on each addition.

Reply:
Absolutely, thanks! 

Related, two lines above we're now summing after expanding to `batch_shape` but _before_ expanding to `batch_size`, which should bring us some extra efficiency. For convenience, here is the previous version

>  logabsdet = self._log_scale.reshape(1, -1).repeat(batch_size, 1).sum(dim=-1)

vs. now (two different locations):
> return self._log_scale.expand(batch_shape).sum() # in `_batch_logabsdet`
> logabsdet = self._batch_logabsdet(batch_shape).expand(batch_size) # in `forward`

Some more easy efficiency gains could come from memoizing `_batch_logabsdet` for different `batch_shape`s, but I'll leave that for now.